### DISCUSSION TITLE
Accounts (and others) not iterable

### DISCUSSION THREAD

**Discussion by newoldtrader** [KmiQ/coinbase-advanced-python#14] [General]
I'm new to python, so there may be a simple solution to this. 

After installing and trying out the example code, I attempted to sort the returned accounts_page.accounts using
`accounts_page.accounts.sort(key=lambda x: x['name'])
`
I kept getting:
TypeError: 'Account' object is not subscriptable

I found that by adding
`    def __getitem__(self, item):
        return getattr(self, item)
`
in the Account class allows the sort to work.

**Comment by KmiQ** [comment]
Thanks for reporting this @newoldtrader.  I agree we can improve the containers like `AccountsPage` to make it organically iterable. I'm going to create an issue for this to see if I can give a shot shortly and include it in the next release. That said, for your specific example, I think you can make that code work if you eliminate the `[]` indexing. The below code should work for you:

`page.accounts.sort(key=lambda x: x.name)` (_Note we are doing `x.name` instead of `x[name]`_).

**Reply by newoldtrader** [reply]
@KmiQ I've been playing around with a few ideas on how to make the containers iterable. Something I've found that solves a bunch of _my_ problems is DotWiz.

```
from dotwiz import DotWiz

class Account(DotWiz):
...

class AccountsPage(DotWiz):
...

```
What this does is essentially turn the containers into dictionaries  that are accessible with both dot notation as well as indexing. It also solves the issue of the exchange adding new responses.
NOTE: some of the class definitions will also need other massaging of the data before insertion. Any datetime types are an example.

As a side benefit:
```
d = client.list_accounts()
for k, v in d.items()
   print(k, v)
pprint(d)
```
also work, making (to me anyway) debugging much easier.

I realize that this is a major change to the code base that may interfere with other requirements, so I'd like your feedback before pursuing things any further.

Thanks

**Reply by KmiQ** [reply]
Thank you for continuing flowing ideas to make the repo better. Some comments below:

**Reply by KmiQ** [reply]
>  It also solves the issue of the exchange adding new responses.

This was already solved in [this PR](https://github.com/KmiQ/coinbase-advanced-python/pull/24). It was part of the `0.1.1` release by the way, so if you update the library you will see it.

**Reply by KmiQ** [reply]
> on how to make the containers iterable.

Coincidentally, I just merged https://github.com/KmiQ/coinbase-advanced-python/pull/27, which makes Page objects iterable.

**Reply by KmiQ** [reply]
> with both dot notation as well as indexing

I'm not sure about the need to have both notations, BUT if you want to access the object fields using indexing, you can always use `__dict__` field. For example: `page.accounts.sort(key=lambda x: x.__dict__['name'])` The way the object are created allow the use of the internal dict.

**Reply by KmiQ** [reply]
Most of this already merged PR should be part of next week's `0.2.0` release.

**Reply by KmiQ** [reply]
So, at this point, I'm not totally sure if all your needs are covered in these latest PRs merged, I'd love to learn more about your use cases to justify using `DotWiz.` I wonder if that's a direction we want to go. We probably need to clarify what extra benefit will add and whether this benefit outweighs the cost (_time invested + making the codebase potentially more complex_). As an opportunity cost, wondering if we are going to benefit more from investing that time in other matters like Rate Limiting and Websocket support.

**Reply by newoldtrader** [reply]
No problem, I'm fine with your solution. Just throwing out ideas.
